Add tests for User

sirkrypt0 opened the pull request:
Closes #9

sirkrypt0 committed:
Add test helper and tests for user model #9

sirkrypt0 committed:
Add postgres service to CI

frcroth commented:
@sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.

sirkrypt0 commented:
> @sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.

We could do something like the following, which should be sufficient imo. PRs should be opened as Draft as soon as work on a new feature has begun, which would make the `push` hook on other branches than the `main` obsolete.

```

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
```

sirkrypt0 commented:
> > @sirkrypt0 Perhaps you can remove the pull_request Hook in the CI in this PR to reduce the number of checks, as it seems unnecessary.
> 
> We could do something like the following, which should be sufficient imo. PRs should be opened as Draft as soon as work on a new feature has begun, which would make the `push` hook on other branches than the `main` obsolete.
> 
> ```
> 
> on:
>   push:
>     branches: [ main ]
>   pull_request:
>     branches: [ main ]
> ```

We might as well allow the CI on all pull requests, not only on main.

frcroth commented:
Yes all PRs and push on main seems like a sensible setting.

sirkrypt0 committed:
Update CI to run on all prs, push only on main

sirkrypt0 commented:
@frcroth I think we can merge this already. I will add route/integration tests in another PR. This allows new MRs to already profit from the test helper and write tests for the model.

frcroth commented on server/src/user/model.rs:
```diff
@@ -65,3 +65,114 @@ impl User {
         Ok(count)
     }
 }
+
```
I'm strongly for moving the tests to another file

sirkrypt0 replied:
Actually, the book states:

>  You’ll put unit tests in the src directory in each file with the code that they’re testing. The convention is to create a module named tests in each file to contain the test functions and to annotate the module with cfg(test).
> https://doc.rust-lang.org/book/ch11-03-test-organization.html#unit-tests

sirkrypt0 replied:
Integration tests will be placed into a `tests` directory

frcroth commented on server/src/user/model.rs:
```diff
@@ -65,3 +65,114 @@ impl User {
+
+    fn create_user_data() -> UserData {
+        UserData {
+            username: String::from("username"),
+            password: String::from("password"),
```
Can we use better standard data?
```suggestion
            username: String::from("testUser"),
            password: String::from("12345678"),
```

Or alternatively use random strings

frcroth commented on server/src/user/model.rs:
```diff
@@ -65,3 +65,114 @@ impl User {
+    }
+
+    #[test]
+    fn destroy_returns_count_if_user_exists() {
```
```suggestion
    fn destroy_returns_one_if_user_exists() {
```

frcroth commented:
Ok I see now that is rust convention to put the tests in the same file. I do not like it but I will accept it.

sirkrypt0 committed:
Use better test data and rename user destroy test

Co-authored-by: frcroth <frcroth@users.noreply.github.com>

sirkrypt0 commented:
> Ok I see now that is rust convention to put the tests in the same file. I do not like it but I will accept it.

Actually I find it reasonable to have the tests in the same file. That way one can use the tests directly to understand what the code is doing and does not have to browse many files to find the tests.

frcroth reviewed (approved):
Requires some changes because the passwords are now saved as hash, other than that I approve

sirkrypt0 committed:
Add tests for generate_password and update tests for password hash

sirkrypt0 commented:
> Requires some changes because the passwords are now saved as hash, other than that I approve

I updated the tests and added a new one for `generate_password`. Please check if you agree with them.
